src/libraries/Common/src/System/Threading/AsyncOverSyncWithIoCancellation.cs

```diff
@@ -0,0 +1,291 @@
+        /// <param name="func">The function to invoke asynchronously.</param>
+        /// <param name="state">The state to pass to the function.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to register with to cancel the synchronous I/O performed by <paramref name="func"/>.</param>
+        /// <returns>A <see cref="ValueTask"/> that represents the asynchronous operation.</returns>
+        /// <remarks>
+        /// The implementation will queue the invocation of the <paramref name="func"/> to the thread pool, with
+        /// the returned <see cref="ValueTask"/> representing its completion.  If the <paramref name="cancellationToken"/> has
+        /// cancellation requested, the implementation will attempt to use CancelSynchronousIo to cancel any I/O being
+        /// performed by the function.
+        /// </remarks>
+        [AsyncMethodBuilder(typeof(PoolingAsyncValueTaskMethodBuilder<>))]
+        public static async ValueTask<TResult> InvokeAsync<TState, TResult>(Func<TState, TResult> func, TState state, CancellationToken cancellationToken)
+        {
+            // Create the work item state object.  This is used to pass around state through various APIs,
+            // while also serving double duty as the work item used to queue the operation to the thread pool.
+            var workItem = new AsyncOverSyncWithIoCancellation();
+
+            // Queue the work to the thread pool.  This is implemented as a custom awaiter that queues the
+            // awaiter itself to the thread pool.
+            await workItem;
+
+            // Register for cancellation, perform the work, and clean up.
+            await using (workItem.RegisterCancellation(cancellationToken).ConfigureAwait(false))
+            {
+                // Even though we're in an async method, awaits _must not_ be used inside
+                // the using block, or else the I/O cancellation could both not work and negatively
+                // interact with I/O on another thread.  The func _must_ be invoked on the same
+                // thread that invoked RegisterCancellation, with no intervening work.
+                try
+                {
+                    return func(state);
+                }
+                catch (OperationCanceledException oce) when (cancellationToken.IsCancellationRequested && oce.CancellationToken != cancellationToken)
+                {
+                    // If the operation fails because of cancellation, make sure it contains this cancellation token
+                    // if this cancellation token could have been the cause.
+                    var newOce = new OperationCanceledException(cancellationToken);
+                    if (oce.StackTrace is string stackTrace)
+                    {
+                        ExceptionDispatchInfo.SetRemoteStackTrace(newOce, stackTrace);
+                    }
```

Comment:
nit: this logic is present in two places, it could be moved to helper method to reduce duplication